### PR TITLE
assertions as array

### LINKED ISSUE #41: Assertion format names
§2.2 Requesting User Information: Editor's note: 

The `assertions` values are lifted from [[RFC8693](https://www.rfc-editor.org/info/rfc8693)]'s "token type identifiers" list, but is there a better source?

§2.4: Identifying the User: Editor's note:

The `assertions` keys are lifted from [RFC8693]'s "token type identifiers" list, but is there a better source? Additionally: should this be an array of objects with internal typing like the sub_ids? Do we expect more than one assertion per user anyway?

§3.4 Returning User Information: Editor's note:

Should `assertion` be an array of objects with internal typing like the sub_ids? Do we expect more than one assertion per user anyway?

### PR DESCRIPTION
**Pull request by fimbault**
Closes #41 with array of objects to account for the possibility of several assertions

### CODE REVIEW

File: draft-ietf-gnap-core-protocol.md [hunk 1]
```diff
@@ -1394,8 +1394,8 @@ sub_ids (array of objects)
 : An array of subject identifiers for the
             end user, as defined by {{I-D.ietf-secevent-subject-identifiers}}.
 
-assertions (object)
-: An object containing assertions as values keyed on the assertion
+assertions (array of objects/strings)
```
Comment (jricher) [inline draft-ietf-gnap-core-protocol.md:1397]:
```suggestion
assertions (array of objects)
```

I don't see anything here about string-formatted assertions, so should this just be objects?

**Review by jricher — CHANGES_REQUESTED** [review]
I think this might need a bit more thought if we're going to change the syntax of it. The old single-object structure was meant to allow for a single ID token, a single SAML assertion, etc. As an array it might make more sense to have the object have an internal "type" field instead, or "format" to match the subject identifier.

**Comment by adeinega** [comment]
As a completely random comment, I find it interesting that computer science commonly refers to an array as a collection of elements with the same data type. It's also a fixed-size collection of elements. https://datatracker.ietf.org/doc/html/rfc8259#section-5 clearly defines and uses an array. Moreover, it says that

> There is no requirement that the values in an array be of the same type.

what's a bit more important is that this specification should probably use "array of ..." everywhere as this isn't true right now. I see the use of both "list of ..." and "array of ..." at the same time.

**Comment by jricher** [comment]
Since GNAP is based around JSON for its data model, we use JSON-native data types and their names. The name for an unlabeled, ordered collection is a "JSON array", as you pointed out. If there are places in GNAP where we're using "list" where we should be saying "array", we should fix that.

File: draft-ietf-gnap-core-protocol.md (hunk 1 above)
Comment (fimbault) [reply draft-ietf-gnap-core-protocol.md:1397]: ok

**Comment by fimbault** [comment]
Todo : 
- inner description to change (key)
- assertions -> format
- subid_formats or something like that

File: draft-ietf-gnap-core-protocol.md [hunk 2]
```diff
@@ -2332,12 +2329,11 @@ sub_ids (array of objects)
             RO, as defined by
             {{I-D.ietf-secevent-subject-identifiers}}.
 
-: An object containing assertions as values keyed on the assertion
+: An array containing assertions as values keyed on the assertion
     type defined by [a registry TBD](#IANA). Possible keys include
     `id_token` for an {{OIDC}} ID Token and `saml2` for a SAML 2 assertion. The assertion
     values are the string serialization of the assertion format, encoded as a plain
     JSON string. Additional assertion types are defined by [a registry TBD](#IANA).
```
Comment (jricher) [inline draft-ietf-gnap-core-protocol.md:2336]:
```suggestion
: An array containing assertions as objects each containing the assertion
    format and the assertion value as the JSON string serialization of the assertion. 
    Possible formats include `id_token` for an {{OIDC}} ID Token and `saml2` for a SAML 2 assertion. 
    Additional assertion formats are defined by [a registry TBD](#IANA).
```

File: draft-ietf-gnap-core-protocol.md [hunk 3]
```diff
@@ -4540,9 +4537,9 @@ key_proofs_supported (array of strings)
...
           [key section](#key-format) of the request.
 
-subject_formats_supported (array of strings)
+sub_ids_formats_supported (array of strings)
 : OPTIONAL. A list of the AS's supported
-          subject identifier types. The values of this list correspond to possible values
+          subject identifier formats. The values of this list correspond to possible values
           of the [subject identifier section](#request-subject) of the request.
 
 assertions_supported (array of strings)
```
Comment (jricher) [inline draft-ietf-gnap-core-protocol.md:4545]:
```suggestion
assertion_formats_supported (array of strings)
```

File: draft-ietf-gnap-core-protocol.md [hunk 4]
```diff
@@ -4540,9 +4537,9 @@ key_proofs_supported (array of strings)
           values of the `proof` field of the
           [key section](#key-format) of the request.
 
-subject_formats_supported (array of strings)
+sub_ids_formats_supported (array of strings)
```
Comment (jricher) [inline draft-ietf-gnap-core-protocol.md:4540]:
```suggestion
sub_id_formats_supported (array of strings)
```

File: draft-ietf-gnap-core-protocol.md [hunk 5]
```diff
@@ -1182,20 +1182,20 @@ the AS, it sends a `subject` field as a JSON object. This object MAY
...
     `id_token` for an {{OIDC}} ID Token and `saml2` for a SAML 2 assertion. Additional
-    assertion values are defined by [a registry TBD](#IANA).
+    assertion formats are defined by [a registry TBD](#IANA).
     \[\[ [See issue #41](https://github.com/ietf-wg-gnap/gnap-core-protocol/issues/41) \]\]
 
 ~~~
 "subject": {
-  "formats": [ "iss_sub", "opaque" ],
-  "assertions": [ "id_token", "saml2" ]
+  "sub_ids_formats": [ "iss_sub", "opaque" ],
```
Comment (jricher) [inline draft-ietf-gnap-core-protocol.md:1197]:
```suggestion
  "sub_id_formats": [ "iss_sub", "opaque" ],
```

File: draft-ietf-gnap-core-protocol.md [hunk 6]
```diff
@@ -1182,20 +1182,20 @@ the AS, it sends a `subject` field as a JSON object. This object MAY
 contain the following fields (or additional fields defined in
 [a registry TBD](#IANA)).
 
-formats (array of strings)
-: An array of subject identifier subject types
+sub_ids_formats (array of strings)
```
Comment (jricher) [inline draft-ietf-gnap-core-protocol.md:1185]:
```suggestion
sub_id_formats (array of strings)
```

File: draft-ietf-gnap-core-protocol.md [hunk 7]
```diff
@@ -988,8 +988,8 @@ A non-normative example of a grant request is below:
         }
     },
     "subject": {
-        "formats": ["iss_sub", "opaque"],
-        "assertions": ["id_token"]
+        "sub_ids_formats": ["iss_sub", "opaque"],
```
Comment (jricher) [inline draft-ietf-gnap-core-protocol.md:991]:
```suggestion
        "sub_id_formats": ["iss_sub", "opaque"],
```

File: draft-ietf-gnap-core-protocol.md [hunk 8]
```diff
@@ -988,8 +988,8 @@ A non-normative example of a grant request is below:
...
+        "assertions_formats": ["id_token"]
```
Comment (jricher) [inline draft-ietf-gnap-core-protocol.md:992]:
```suggestion
        "assertion_formats": ["id_token"]
```

File: draft-ietf-gnap-core-protocol.md [hunk 9]
```diff
@@ -1182,20 +1182,20 @@ the AS, it sends a `subject` field as a JSON object. This object MAY
...
 [a registry TBD](#IANA)).
 
-formats (array of strings)
-: An array of subject identifier subject types
+sub_ids_formats (array of strings)
+: An array of subject identifier subject formats
             requested for the RO, as defined by {{I-D.ietf-secevent-subject-identifiers}}.
 
-assertions (array of strings)
+assertions_formats (array of strings)
```
Comment (jricher) [inline draft-ietf-gnap-core-protocol.md:1189]:
```suggestion
assertion_formats (array of strings)
```

File: draft-ietf-gnap-core-protocol.md [hunk 10]
```diff
@@ -1182,20 +1182,20 @@ the AS, it sends a `subject` field as a JSON object. This object MAY
...
-    assertion values are defined by [a registry TBD](#IANA).
+    assertion formats are defined by [a registry TBD](#IANA).
     \[\[ [See issue #41](https://github.com/ietf-wg-gnap/gnap-core-protocol/issues/41) \]\]
 
 ~~~
 "subject": {
-  "formats": [ "iss_sub", "opaque" ],
-  "assertions": [ "id_token", "saml2" ]
+  "sub_ids_formats": [ "iss_sub", "opaque" ],
+  "assertions_formats": [ "id_token", "saml2" ]
```
Comment (jricher) [inline draft-ietf-gnap-core-protocol.md:1198]:
```suggestion
  "assertion_formats": [ "id_token", "saml2" ]
```

File: draft-ietf-gnap-core-protocol.md [hunk 11]
```diff
@@ -1211,7 +1211,7 @@ channels by the client instance, as discussed in {{response-subject}}.
 
 The AS SHOULD NOT re-use subject identifiers for multiple different ROs.
 
-Note: the "formats" and "assertions" request fields are independent of
+Note: the "sub_ids_formats" and "assertions_formats" request fields are independent of
```
Comment (jricher) [inline draft-ietf-gnap-core-protocol.md:1214]:
```suggestion
Note: the "sub_id_formats" and "assertion_formats" request fields are independent of
```

**Pull request merged**